data/transactions/logic/assembler.go

```diff
@@ -1666,6 +1681,80 @@ func (ops *OpStream) assemble(text string) error {
 	return nil
 }
 
+func (ops *OpStream) cycle(macro string, previous ...string) bool {
+	replacement, ok := ops.macros[macro]
+	if !ok {
+		return false
+	}
+	if len(previous) > 0 && macro == previous[0] {
+		ops.errorf("Macro cycle discovered: %s", strings.Join(append(previous, macro), " -> "))
+		return true
+	}
+	for _, token := range replacement {
+		if ops.cycle(token, append(previous, macro)...) {
+			return true
+		}
+	}
+	return false
+}
+
+// Once a version is obtained, versionedMacroFilter goes through previously defined macros
+// and ensures they don't use opcodes/fields from the obtained version
+func (ops *OpStream) versionedMacroFilter() error {
+	var err error = nil
+	for macroName := range ops.macros {
+		if fieldNames[ops.Version][macroName] {
+			err = fmt.Errorf("%s is defined as a macro but is a field name in version %d", macroName, ops.Version)
+		} else if _, ok := OpsByName[ops.Version][macroName]; ok {
+			err = fmt.Errorf("%s is defined as a macro but is an opcode in version %d", macroName, ops.Version)
+		} else {
+			continue
+		}
+		ops.error(err)
+		delete(ops.macros, macroName)
+	}
+	if err != nil {
+		err = errors.New("version is incompatible with defined macros")
+	}
+	return err
+}
+
+func (ops *OpStream) filterMacro(macroName string) error {
+	_, isTxnType := txnTypeMap[macroName]
+	_, isOnCompletion := onCompletionMap[macroName]
+	if isTxnType || isOnCompletion {
+		return ops.errorf("Named constants cannot be used as macro names: %s", macroName)
```

Comment:
Is it currently possible for macros to be numbers? E.g.
```
#define 0 123
int 0 // actually int 123
```

If so, that seems undesirable and should probably be filtered out as well

Reply:
Good point! Numbers and labels will be added to the filter soon

Reply:
Great point, and given the very "lenient" way we think of tokens in the assembler, this passes.
```
testProg(t, `
     #define "john" 1
     int "john"
`, AssemblerMaxVersion)
```

Reply:
I will say I don't really see an easy way of stopping this without either making a bunch of special rules in define or changing our tokens to carry more info than just a string

Reply:
I think we can decide on what a legal macroname is. I vote for:
It consists of alphanumeric characters, plus (at least) the characters that we have already used in opcodes: +,-,%,*,/, etc.  It cannot start with a digit.

Reply:
Seems reasonable to me; I'm a little worried there's still going to be some edge cases, but I guess that's not the worst thing. For example, another rule we will likely want is if the name begins with a hyphen, then the second char cannot be a digit

Reply:
Oh. Don't forget to ban redefining pseudo-ops, not just opcodes.